# **1.a**

In [1]:
import torch
import torch.nn as nn
import torch.optim as optim
import torchvision
import torchvision.transforms as transforms

# Define the neural network
class SimpleNN(nn.Module):
    def __init__(self, input_size, hidden_size, num_classes):
        super(SimpleNN, self).__init__()
        self.fc1 = nn.Linear(input_size, hidden_size)
        self.relu = nn.ReLU()
        self.fc2 = nn.Linear(hidden_size, num_classes)

    def forward(self, x):
        x = self.fc1(x)
        x = self.relu(x)
        x = self.fc2(x)
        return x

# CIFAR-10 dataset
transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
])

train_dataset = torchvision.datasets.CIFAR10(root='./data', train=True, download=True, transform=transform)
train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=64, shuffle=True)

# Hyperparameters
input_size = 3 * 32 * 32  # CIFAR-10 images are 32x32x3
hidden_size = 512
num_classes = 10
learning_rate = 0.001
num_epochs = 5

# Initialize the model
model = SimpleNN(input_size, hidden_size, num_classes)

# Loss and optimizer
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr=learning_rate)

# Training loop
total_step = len(train_loader)
for epoch in range(num_epochs):
    for i, (images, labels) in enumerate(train_loader):
        # Reshape images to (batch_size, input_size)
        images = images.reshape(-1, input_size)

        # Forward pass
        outputs = model(images)
        loss = criterion(outputs, labels)

        # Backward pass and optimize
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        if (i+1) % 100 == 0:
            print ('Epoch [{}/{}], Step [{}/{}], Loss: {:.4f}'
                   .format(epoch+1, num_epochs, i+1, total_step, loss.item()))

# Save the model checkpoint
torch.save(model.state_dict(), 'simple_nn.ckpt')

100%|██████████| 170498071/170498071 [00:03<00:00, 43242474.84it/s]


Extracting ./data/cifar-10-python.tar.gz to ./data
Epoch [1/5], Step [100/782], Loss: 2.2687
Epoch [1/5], Step [200/782], Loss: 2.2014
Epoch [1/5], Step [300/782], Loss: 2.2348
Epoch [1/5], Step [400/782], Loss: 2.0980
Epoch [1/5], Step [500/782], Loss: 2.1399
Epoch [1/5], Step [600/782], Loss: 2.1114
Epoch [1/5], Step [700/782], Loss: 2.0175
Epoch [2/5], Step [100/782], Loss: 2.0585
Epoch [2/5], Step [200/782], Loss: 2.0730
Epoch [2/5], Step [300/782], Loss: 2.0765
Epoch [2/5], Step [400/782], Loss: 2.0241
Epoch [2/5], Step [500/782], Loss: 2.1622
Epoch [2/5], Step [600/782], Loss: 2.0549
Epoch [2/5], Step [700/782], Loss: 1.9072
Epoch [3/5], Step [100/782], Loss: 1.9467
Epoch [3/5], Step [200/782], Loss: 1.9146
Epoch [3/5], Step [300/782], Loss: 1.9513
Epoch [3/5], Step [400/782], Loss: 1.9674
Epoch [3/5], Step [500/782], Loss: 1.9518
Epoch [3/5], Step [600/782], Loss: 1.8743
Epoch [3/5], Step [700/782], Loss: 1.9520
Epoch [4/5], Step [100/782], Loss: 1.9006
Epoch [4/5], Step [200/78

# **1.b**

In [4]:

# Define the neural network with additional hidden layers
class ExtendedNN(nn.Module):
    def __init__(self, input_size, hidden_size1, hidden_size2, hidden_size3, num_classes):
        super(ExtendedNN, self).__init__()
        self.fc1 = nn.Linear(input_size, hidden_size1)
        self.relu1 = nn.ReLU()
        self.fc2 = nn.Linear(hidden_size1, hidden_size2)
        self.relu2 = nn.ReLU()
        self.fc3 = nn.Linear(hidden_size2, hidden_size3)
        self.relu3 = nn.ReLU()
        self.fc4 = nn.Linear(hidden_size3, num_classes)

    def forward(self, x):
        x = self.fc1(x)
        x = self.relu1(x)
        x = self.fc2(x)
        x = self.relu2(x)
        x = self.fc3(x)
        x = self.relu3(x)
        x = self.fc4(x)
        return x


In [5]:
# CIFAR-10 dataset
transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
])

train_dataset = torchvision.datasets.CIFAR10(root='./data', train=True, download=True, transform=transform)
train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=64, shuffle=True)


Files already downloaded and verified


In [6]:
# Hyperparameters
input_size = 3 * 32 * 32  # CIFAR-10 images are 32x32x3
hidden_size1 = 512
hidden_size2 = 256
hidden_size3 = 128
num_classes = 10
learning_rate = 0.001
num_epochs = 300


In [7]:
# Initialize the model
model = ExtendedNN(input_size, hidden_size1, hidden_size2, hidden_size3, num_classes)

# Loss and optimizer
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr=learning_rate)


In [9]:
# Training loop

total_step = len(train_loader)
for epoch in range(num_epochs):
    for i, (images, labels) in enumerate(train_loader):
        # Reshape images to (batch_size, input_size)
        images = images.reshape(-1, input_size)

        # Forward pass
        outputs = model(images)
        loss = criterion(outputs, labels)

        # Backward pass and optimize
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        if (i+1) % 100 == 0:
            print ('Epoch [{}/{}], Step [{}/{}], Loss: {:.4f}'
                   .format(epoch+1, num_epochs, i+1, total_step, loss.item()))



Epoch [1/300], Step [100/782], Loss: 2.3176
Epoch [1/300], Step [200/782], Loss: 2.3081
Epoch [1/300], Step [300/782], Loss: 2.2963
Epoch [1/300], Step [400/782], Loss: 2.2939
Epoch [1/300], Step [500/782], Loss: 2.2896
Epoch [1/300], Step [600/782], Loss: 2.3007
Epoch [1/300], Step [700/782], Loss: 2.3049
Epoch [2/300], Step [100/782], Loss: 2.2920
Epoch [2/300], Step [200/782], Loss: 2.2980
Epoch [2/300], Step [300/782], Loss: 2.2910
Epoch [2/300], Step [400/782], Loss: 2.2873
Epoch [2/300], Step [500/782], Loss: 2.2974
Epoch [2/300], Step [600/782], Loss: 2.2912
Epoch [2/300], Step [700/782], Loss: 2.2770
Epoch [3/300], Step [100/782], Loss: 2.2779
Epoch [3/300], Step [200/782], Loss: 2.2784
Epoch [3/300], Step [300/782], Loss: 2.2714
Epoch [3/300], Step [400/782], Loss: 2.2729
Epoch [3/300], Step [500/782], Loss: 2.2777
Epoch [3/300], Step [600/782], Loss: 2.2664
Epoch [3/300], Step [700/782], Loss: 2.2750
Epoch [4/300], Step [100/782], Loss: 2.2806
Epoch [4/300], Step [200/782], L

In [10]:
# Evaluation
model.eval()
with torch.no_grad():
    correct = 0
    total = 0
    for images, labels in train_loader:
        images = images.reshape(-1, input_size)
        outputs = model(images)
        _, predicted = torch.max(outputs.data, 1)
        total += labels.size(0)
        correct += (predicted == labels).sum().item()

    print('Accuracy on the {} training images: {} %'.format(total, 100 * correct / total))

# Save the model checkpoint
torch.save(model.state_dict(), 'extended_nn.ckpt')

Accuracy on the 50000 training images: 99.696 %


Loss for 1a. = 1.8256

Loss for 1b. = 0.0566

Acurracy for 3 hidden layer = 99.696%


Data set fits into the 3hidden layer model well with low losses and high Accuracy.

